dds_web/api/schemas/file_schemas.py

```diff
@@ -0,0 +1,93 @@
+"""File related marshmallow schemas."""
+
+####################################################################################################
+# IMPORTS ################################################################################ IMPORTS #
+####################################################################################################
+
+# Standard Library
+from datetime import datetime
+
+# Installed
+import marshmallow
+import sqlalchemy
+
+# Own modules
+from dds_web.database import models
+import dds_web.utils
+from dds_web.api.schemas import project_schemas
+
+####################################################################################################
+# SCHEMAS ################################################################################ SCHEMAS #
+####################################################################################################
+
+
+class NewFileSchema(project_schemas.ProjectRequiredSchema):
+    """Validates and creates a new file object."""
+
+    # Length minimum 1 required, required=True accepts empty string
+    name = marshmallow.fields.String(required=True, validate=marshmallow.validate.Length(min=1))
+    name_in_bucket = marshmallow.fields.String(
+        required=True, validate=marshmallow.validate.Length(min=1)
+    )
+    subpath = marshmallow.fields.String(required=True, validate=marshmallow.validate.Length(min=1))
+    size = marshmallow.fields.Integer(required=True)  # Accepts BigInt
+    size_processed = marshmallow.fields.Integer(required=True)  # Accepts BigInt
+    compressed = marshmallow.fields.Boolean(required=True)  # Accepts all truthy
+    public_key = marshmallow.fields.String(
+        required=True, validate=marshmallow.validate.Length(equal=64)
+    )
+    salt = marshmallow.fields.String(required=True, validate=marshmallow.validate.Length(equal=32))
+    checksum = marshmallow.fields.String(
+        required=True, validate=marshmallow.validate.Length(equal=64)
+    )
+
+    @marshmallow.validates_schema(skip_on_field_errors=True)
+    def verify_file_not_exists(self, data, **kwargs):
+        """Check that the file does not match anything already in the database."""
+
+        # Check that there is no such file in the database
+        project = data.get("project_row")
+        try:
+            file = (
+                models.File.query.filter(
+                    sqlalchemy.and_(
+                        models.File.name == sqlalchemy.func.binary(data.get("name")),
+                        models.File.project_id == sqlalchemy.func.binary(project.id),
+                    )
+                )
+                .with_entities(models.File.id)
+                .one_or_none()
+            )
+        except sqlalchemy.exc.SQLAlchemyError:
+            raise
+
+        if file:
+            raise FileExistsError
```

Comment:
Just for me to understand: The built-in OSerror `FileExistsError` is raised here, because raising a custom error with a respective API response would leak information?

Reply:
No just trying to use as many non-custom exceptions as possible. In this case this really shouldn't happen so it should completely fail. Could also use ValidationError which will fail as well, I just added FileExistsError because it seemed appropriate at the time. 👍🏻